sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1774,6 +1775,51 @@ class Analyzer(
   }
 
   /**
+   * Substitute Hints.
+   * - BROADCAST/BROADCASTJOIN/MAPJOIN match the closest table with the given name parameters.
+   */
+  object SubstituteHints extends Rule[LogicalPlan] {
+    def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+      case logical: LogicalPlan => logical transformDown {
+        case h @ Hint(name, parameters, child)
+            if Seq("BROADCAST", "BROADCASTJOIN", "MAPJOIN").contains(name.toUpperCase) =>
+          var resolvedChild = child
+
+          for (param <- parameters) {
+            val names = param.split("\\.")
+            val tid = if (names.length > 1) {
+              TableIdentifier(names(1), Some(names(0)))
+            } else {
+              TableIdentifier(param, None)
+            }
+            try {
+              catalog.lookupRelation(tid)
+
+              var stop = false
+              resolvedChild = resolvedChild.transformDown {
+                case r @ BroadcastHint(SubqueryAlias(t, _))
+                  if !stop && resolver(t, tid.identifier) =>
+                  stop = true
+                  r
+                case r @ SubqueryAlias(t, _) if !stop && resolver(t, tid.identifier) =>
+                  stop = true
+                  BroadcastHint(r)
```

Comment:
When `BroadcastHint` is crossing `SubqueryAlias`, we need to remove the qualifier.